File: 20-Valid-Parentheses.cpp
class Solution {
public:
    bool isValid(string str) {
        unordered_map<char, char> brackets = {
            {'[', ']'},
            {'(', ')'},
            {'{', '}'}
        };

        stack<char> stk;

        for (char c: str) {
            if (brackets.count(c)) stk.push(c);
            else {
                if (stk.empty() || brackets[stk.top()] != c) return false;

                stk.pop();
            }
        }

        return stk.empty();
        
    }
};

File: 20-Valid-Parentheses.java
class Solution {
    public boolean isValid(String s) {
        Map<Character, Character> parantheses = new HashMap<>();
        parantheses.put(')', '(');
        parantheses.put('}', '{');
        parantheses.put(']', '[');

        Stack<Character> stk = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stk.push(c);
            } else {
                if (stk.isEmpty() || stk.pop() != parantheses.get(c)) {
                    return false;
                }
            }
        }
        return stk.size() == 0;
    }
}